### PR TITLE
P2P Bluetooth peer discovery changes

### PR DESCRIPTION
**Pull request by Aniket392**
- Adding c4Peerdiscovery implementation, implementing all methods and connecting it to java
- BT connections discovery logic from both native as well as java side
- Writing helper class like MetadataHelper, to convert java map object
- Adding native implementation to BluetoothPeer
- Updating native_c4multipeerreplicator to take protocol parameter

### CODE REVIEW

File: android/main/AndroidManifest.xml [hunk 1]
```diff
@@ -20,6 +20,7 @@ limitations under the License.
     xmlns:tools="http://schemas.android.com/tools"
     >
 
+    <uses-permission android:name="android.permission.BLUETOOTH_CONNECT" />
```
Comment (pasin) [inline android/main/AndroidManifest.xml]: What does this mean to users who don't use MultipeerReplicator with Bluetooth?

File: common/main/cpp/native_c4multipeerreplicator.cc [hunk 2]
```diff
@@ -525,9 +528,7 @@ JNICALL Java_com_couchbase_lite_internal_core_impl_NativeC4MultipeerReplicator_c
     params.peerGroupID = groupId;
 
     // Protocols:
-    C4String protocols[] = {kPeerSyncProtocol_DNS_SD};
-    params.protocols = protocols;
-    params.protocolsCount = sizeof(protocols) / sizeof(protocols[0]);
+    params.protocols = kPeerSyncProtocol_BluetoothLE;
```
Comment (pasin) [inline common/main/cpp/native_c4multipeerreplicator.cc]: Need to be passed as an array of transport values.

File: common/main/cpp/native_c4peerdiscoveryprovider.cc [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+                stopPublishing();
+            }
+        }
+
+        void addDiscoveredPeer(C4Peer* peer, bool moreComing = false) {
+            addPeer(peer, moreComing);
+        }
+
+        void removeDiscoveredPeer(std::string id, bool moreComing = false) {
+            removePeer(id, moreComing);
```
Comment (pasin) [inline common/main/cpp/native_c4peerdiscoveryprovider.cc]: Cannot use `removePeer(id, moreComing)` directly?

File: common/main/cpp/native_c4peerdiscoveryprovider.cc [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+        const char* peerIdChars = env->GetStringUTFChars(peerId, nullptr);
+        if (peerIdChars) {
+            peerIdStr = std::string(peerIdChars);
+            env->ReleaseStringUTFChars(peerId, peerIdChars);
+        }
+    }
+
+    // === CALL LIBRARY API ===
+    C4PeerDiscovery& discovery = provider->discovery();
+    fleece::Retained<C4Peer> peer = discovery.peerWithID(peerIdStr);
```
Comment (pasin) [inline common/main/cpp/native_c4peerdiscoveryprovider.cc]: This peer is a Retained object and its ref count will be decreased by 1 after this function is returned (done). In line 456, its raw pointer is being passed to the Java object and as I understand, the Java object will keep the raw pointer to use later. 

The problem with this is that the peer's raw pointer could be freed any time depending on the logic inside the discovery. It's unsafe to use by the Java object that keeps the raw pointer and the crash is due to happen. 

Therefore, it's better to keep the raw pointer with ref count + 1 by calling peer.detach() to get that. Then pass the pointer with ref count + 1 to the Java object instead. However, when Java object is done, it will need to call to release the pointer as well. 

To implement that, you can add a C function such as C4Peer_Release(C4Peer* peer) for Java to called via the JNI. Then the Java class needs to implement C4Peer if I understand. I'm not sure about the Cleaner stuff but I think you can follow some of the existing classes that do that.

File: common/main/cpp/native_c4peerdiscoveryprovider.cc [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+            env->ReleaseByteArrayElements(peerId, bytes, JNI_ABORT);
+        }
+    }
+
+    // Get the C4Socket from the pointer
+    C4Socket* socket = (C4Socket*)socketPtr;
+    if (!socket) return;
+
+    // Find the peer in the discovery system
+    auto peer = provider->discovery().peerWithID("peerIdObj");
```
Comment (pasin) [inline common/main/cpp/native_c4peerdiscoveryprovider.cc]: Is `peerWithID("peerIdObj")` correct with the "peerIdObject" string?

**Review by pasin — CHANGES_REQUESTED** [review]

**Review by pasin — CHANGES_REQUESTED** [review]

File: common/main/cpp/native_c4multipeerreplicator.cc [hunk 6]
```diff
@@ -688,7 +689,7 @@ JNICALL Java_com_couchbase_lite_internal_core_impl_NativeC4MultipeerReplicator_g
             m_C4MultipeerReplicator_createPeerInfo,
             jpeerId,
             certChain,
-            info->online ? JNI_TRUE : JNI_FALSE,
+            JNI_TRUE,
```
[the remark this answers is not in the document]
Comment (Aniket392) [reply common/main/cpp/native_c4multipeerreplicator.cc]: Updated as shared

File: common/main/cpp/native_c4peerdiscoveryprovider.cc [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+            if ((envState != JNI_OK) && (envState != JNI_EDETACHED))
+                return;
+
+            // Call Java BLE service to start scanning
+            env->CallStaticVoidMethod(
+                    cls_C4PeerDiscoveryProvider,
+                    m_C4PeerDiscoveryProvider_startBrowsing,
+                    _contextToken);
+
+
```
[the remark this answers is not in the document]
Comment (Aniket392) [reply common/main/cpp/native_c4peerdiscoveryprovider.cc]: Removed extra spaces

File: common/main/cpp/native_c4peerdiscoveryprovider.cc [hunk 8]
```diff
@@ -0,0 +1,583 @@
...
+                    m_C4PeerDiscoveryProvider_stopPublishing,
+                    _contextToken);
+
+            if (envState == JNI_EDETACHED) {
+                detachJVM("stopPublishing");
+            }
+        }
+
+
+
```
[the remark this answers is not in the document]
Comment (Aniket392) [reply common/main/cpp/native_c4peerdiscoveryprovider.cc]: Removed extra lines

File: common/main/cpp/native_c4peerdiscoveryprovider.cc [hunk 9]
```diff
@@ -0,0 +1,583 @@
...
+    // Get the C4Socket from the pointer
+    C4Socket* socket = (C4Socket*)socketPtr;
+    if (!socket) return;
+
+    // Find the peer in the discovery system
+    auto peer = provider->discovery().peerWithID("peerIdObj");
+    if (peer) {
+        // Notify about the incoming connection
+        bool accepted = provider->notifyIncomConnection(peer.get(), socket);
```
[the remark this answers is not in the document]
Comment (Aniket392) [reply common/main/cpp/native_c4peerdiscoveryprovider.cc]: Currently removed this
Will update with Socketfactory

**Review by pasin — CHANGES_REQUESTED** [review]

**Review by pasin — CHANGES_REQUESTED** [review]

**Review by pasin — APPROVED** [review]
